pkg/extensions/sync/service.go

```diff
@@ -90,100 +93,105 @@ func New(
-		return nil
+	if urls[0].Scheme == "http" {
+		tls = config.TLSDisabled
 	}
 
-	found := false
-
-	for _, url := range service.config.URLs {
-		// skip current client
-		if service.client != nil && service.client.GetBaseURL() == url {
-			continue
-		}
+	mirrorsHosts := make([]string, 0)
+	for _, url := range urls[1:] {
+		mirrorsHosts = append(mirrorsHosts, url.Host)
+	}
 
-		remoteAddress := StripRegistryTransport(url)
-		credentials := service.credentials[remoteAddress]
+	hostConfig := config.Host{}
+	hostConfig.Name = mainHost
+	hostConfig.Mirrors = mirrorsHosts
+	hostConfig.TLS = tls
 
-		tlsVerify := true
-		if service.config.TLSVerify != nil {
-			tlsVerify = *service.config.TLSVerify
+	if opts.CertDir != "" {
+		clientCert, clientKey, regCert, err := getCertificates(opts.CertDir)
+		if err != nil {
+			return nil, err
 		}
 
-		options := client.Config{
-			URL:       url,
-			Username:  credentials.Username,
-			Password:  credentials.Password,
-			TLSVerify: tlsVerify,
-			CertDir:   service.config.CertDir,
-		}
+		hostConfig.ClientCert = clientCert
+		hostConfig.ClientKey = clientKey
+		hostConfig.RegCert = regCert
+	}
 
-		var err error
+	if mainHost == regclient.DockerRegistryAuth ||
+		mainHost == regclient.DockerRegistryDNS ||
+		mainHost == regclient.DockerRegistry ||
+		mainHost == "index.docker.io" {
+		hostConfig.Name = regclient.DockerRegistry
+		hostConfig.Hostname = regclient.DockerRegistryDNS
+		hostConfig.CredHost = regclient.DockerRegistryAuth
+	}
 
-		if service.client != nil {
-			err = service.client.SetConfig(options)
-		} else {
-			service.client, err = client.New(options, service.log)
-		}
+	credentials, ok := credentialsFile[mainHost]
+	if ok {
+		hostConfig.User = credentials.Username
+		hostConfig.Pass = credentials.Password
+	}
 
-		if err != nil {
-			service.log.Error().Err(err).Str("url", url).Msg("failed to initialize http client")
+	hostConfigOpts := []config.Host{}
+	hostConfigOpts = append(hostConfigOpts, hostConfig)
 
-			return err
-		}
+	for _, mirror := range mirrorsHosts {
+		mirroHostConfig := hostConfig
+		mirroHostConfig.Name = mirror
+		hostConfigOpts = append(hostConfigOpts, mirroHostConfig)
+	}
 
-		if service.client.Ping() {
-			found = true
+	service.regclient = regclient.New(
+		regclient.WithDockerCerts(),
+		regclient.WithDockerCreds(),
+		regclient.WithRegOpts(
+			reg.WithCertDirs([]string{opts.CertDir}),
+			reg.WithRetryLimit(maxRetries),
+			reg.WithDelay(1*time.Second, retryDelay),
+		),
+		regclient.WithConfigHost(hostConfigOpts...),
+	)
 
-			break
-		}
-	}
+	hosts := []string{}
+	hosts = append(hosts, mainHost)
+	hosts = append(hosts, mirrorsHosts...)
 
-	if service.client == nil || !found {
-		return zerr.ErrSyncPingRegistry
-	}
+	service.remote = NewRemoteRegistry(
+		service.regclient,
+		hosts,
+		service.log,
+	)
 
-	return nil
-}
+	// we want referrers using sha-<digest>.*" tags
+	// service.copyOptions = append(service.copyOptions, regclient.ImageWithDigestTags())
```

Comment:
If I let both options enabled, regclient will overwrite some of the referrers.
Because regclient will create tags for all referrers it copies, even if there's no referrer tag on upstream.

I think this issue arises only if the upstream is older zot, or the upstream supports both digest referrers and subject referrers.

I'm not 100% percent sure what is the exact behaviour though.

I will try to fully grasp it.

Reply:
You mean zot manages the referrers tag internally, and updates it every time we copy something to the storage? In that case I think we should not copy it from upstream.

Reply:
I just want to make sure this PR resolves https://github.com/project-zot/zot/issues/2417